### PR TITLE
Upgrade to Kafka version V2.0.0 and Improve Documentation

### LINKED ISSUE #21: Record type name `Offset` is misleading
**Description:**
> `Package-Kafka` has a record type of `Offset` which stores the offset value and the `TopicPartition` record values. This name is misleading as the `Offset` itself contains a field called `offset`, where the actual offset value is stored. This should be renamed to either `TopicOffset` or `TopicPartitionOffset`.

**Suggested Labels:**
> Enhancement

**Suggested Assignees:**
> @ThisaruGuruge 

**Affected Product Version:**
> 0.971.0

### PR DESCRIPTION
**Pull request by ThisaruGuruge**
## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -370,10 +370,10 @@
     <properties>
         <scala.version>2.11.0</scala.version>
         <version.kafka.scala>2.11</version.kafka.scala>
-        <debezium.version>0.7.0</debezium.version>
+        <debezium.version>0.9.0.Alpha2</debezium.version>
         <project.scm.id>my-scm-server</project.scm.id>
-        <ballerina.version>0.982.0</ballerina.version>
-        <apache.kafka.version>1.0.0</apache.kafka.version>
+        <ballerina.version>0.982.1-SNAPSHOT</ballerina.version>
```
[the remark this answers is not in the document]
Comment (ThisaruGuruge) [reply pom.xml]: This was checked with Ballerina 0.982.1-SNAPSHOT.
As this is going with the next release. Now the ballerina version is changed back to 0.982.0, will change the version when releasing.

File: component/src/main/java/org/ballerinalang/kafka/nativeimpl/consumer/action/AbstractApisWithDuration.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (Anoukh) [inline component/src/main/java/org/ballerinalang/kafka/nativeimpl/consumer/action/AbstractApisWithDuration.java]: Should this be 2018?

File: component/src/main/java/org/ballerinalang/kafka/nativeimpl/consumer/action/AbstractApisWithDuration.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+            duration = getDefaultApiTimeoutConsumerConfig(consumerProperties);
+        } else {
+            duration = -1;
+        }
+        return duration;
+    }
+
+    protected int getDefaultApiTimeoutConsumerConfig(Properties consumerProperties) {
+        int apiTimeoutValue = (int) consumerProperties.get(ConsumerConfig.DEFAULT_API_TIMEOUT_MS_CONFIG);
+        return apiTimeoutValue;
```
Comment (Anoukh) [inline component/src/main/java/org/ballerinalang/kafka/nativeimpl/consumer/action/AbstractApisWithDuration.java]: Can we merge with above line and avoid the additional variable?

File: component/src/main/java/org/ballerinalang/kafka/nativeimpl/consumer/action/GetAvailableTopics.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
+package org.ballerinalang.kafka.nativeimpl.consumer.action;
```
Comment (Anoukh) [inline component/src/main/java/org/ballerinalang/kafka/nativeimpl/consumer/action/GetAvailableTopics.java]: Missing licence header?

File: component/src/main/java/org/ballerinalang/kafka/nativeimpl/producer/action/AbortTransaction.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+        String connectorKey = producerConnector.get("connectorID").stringValue();
+        this.producer = (KafkaProducer) producerStruct.getNativeData(NATIVE_PRODUCER);
+        Properties producerProperties = (Properties) producerStruct.getNativeData(NATIVE_PRODUCER_CONFIG);
+
+        LocalTransactionInfo localTransactionInfo = context.getLocalTransactionInfo();
+
+        try {
+            if (isTransactionalProducer(producerProperties)) {
+                if (isKafkaTransactionInitiated(localTransactionInfo, connectorKey)) {
```
Comment (Anoukh) [inline component/src/main/java/org/ballerinalang/kafka/nativeimpl/producer/action/AbortTransaction.java]: Merge the two if conditions with && ?

File: component/src/main/wso2/kafka/consumer.bal [hunk 6]
```diff
@@ -16,43 +16,44 @@
...
-# + maxPollInterval - maximum delay between invocations of poll
-# + reconnectBackoff - time to wait before attempting to reconnect
-# + pollingTimeout - time out interval of polling
-# + pollingInterval - polling interval
-# + concurrentConsumers - number of concurrent consumers
-# + autoCommit - enables auto commit offsets
-# + checkCRCS - check the CRC32 of the records consumed
-# + excludeInternalTopics - whether records from internal topics should be exposed to the consumer
-# + decoupleProcessing - decouples processing
+# + bootstrapServers - List of remote server endpoints
```
Comment (pubudu91) [inline component/src/main/wso2/kafka/consumer.bal]: Shall we be a bit more descriptive than this on what these servers are for?

File: component/src/main/wso2/kafka/consumer.bal [hunk 7]
```diff
@@ -16,43 +16,44 @@
...
+# + groupId - Unique string that identifies the consumer
+# + offsetReset - Offset reset strategy if no initial offset
+# + partitionAssignmentStrategy - Strategy class for handle partition assignment among consumers
```
Comment (pubudu91) [inline component/src/main/wso2/kafka/consumer.bal]: handle -> handling

File: component/src/main/wso2/kafka/consumer.bal [hunk 8]
```diff
@@ -16,43 +16,44 @@
...
+# + metricsRecordingLevel - Metrics recording level
+# + metricsReporterClasses - Metrics reporter classes
+# + clientId - Id to be used for server side logging
```
Comment (pubudu91) [inline component/src/main/wso2/kafka/consumer.bal]: Id -> ID

File: component/src/main/wso2/kafka/consumer.bal [hunk 9]
```diff
@@ -16,43 +16,44 @@
...
+# + interceptorClasses - Interceptor classes to be used before sending records
+# + isolationLevel - How the transactional messages are read
```
Comment (pubudu91) [inline component/src/main/wso2/kafka/consumer.bal]: Shall we rephrase this such that the association between the isolationLevel and the messages is clear?

File: component/src/main/wso2/kafka/consumer.bal [hunk 10]
```diff
@@ -16,43 +16,44 @@
...
+# + topics - Topics to be subscribed
+# + properties - Additional properties if required
+# + sessionTimeout - Timeout used to detect consumer failures when heartbeat threshold is reached
+# + heartBeatInterval - Expected time between heartbeats
+# + metadataMaxAge - Max time to force a refresh of metadata
+# + autoCommitInterval - Offset committing interval
```
Comment (pubudu91) [inline component/src/main/wso2/kafka/consumer.bal]: Shall we rephrase this? Currently this comes out as a sentence fragment

File: component/src/main/wso2/kafka/consumer.bal [hunk 11]
```diff
@@ -16,43 +16,44 @@
...
+# + maxPartitionFetchBytes - The max amount of data per-partition the server return
```
Comment (pubudu91) [inline component/src/main/wso2/kafka/consumer.bal]: return -> returns

Comment (pubudu91) [reply component/src/main/wso2/kafka/consumer.bal]: Also, shall we consistently use either max or maximum (better if maximum is used)? In some places it says maximum.

File: component/src/main/wso2/kafka/consumer.bal [hunk 12]
```diff
@@ -99,12 +101,12 @@ public type ConsumerConfig record {
 
 # Type related to consumer record.
 #
-# + key - key that is included in the record
-# + value - record content
-# + offset - offset value
-# + partition - partition to which the record is stored
-# + timestamp - timestamp of the record, in milliseconds since epoch
-# + topic - topic of the record
+# + key - Key that is included in the record.
```
Comment (pubudu91) [inline component/src/main/wso2/kafka/consumer.bal]: Again, need to be consistent. Here, the sentences are finished with a period while in the doc block above, that is not the case

File: component/src/main/wso2/kafka/consumer.bal [hunk 13]
```diff
@@ -99,12 +101,12 @@ public type ConsumerConfig record {
...
+# + value - Record content.
+# + offset - Offset value.
+# + partition - Partition to which the record is stored.
```
Comment (pubudu91) [inline component/src/main/wso2/kafka/consumer.bal]: Partition to -> Partition in

File: component/src/main/wso2/kafka/consumer.bal [hunk 14]
```diff
@@ -129,16 +131,16 @@ public type Consumer object {
 
 # Represent a Kafka consumer endpoint.
 #
-# + consumerActions - handle all the actions related to the endpoint
-# + consumerConfig - used to store configurations related to a Kafka connection
+# + consumerActions - Handle all the actions related to the endpoint.
```
Comment (pubudu91) [inline component/src/main/wso2/kafka/consumer.bal]: Handle -> Handles

File: component/src/main/java/org/ballerinalang/kafka/nativeimpl/producer/action/AbstractTransactionHandler.java [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+public abstract class AbstractTransactionHandler implements NativeCallableUnit {
+
+    protected Context context;
+    protected KafkaProducer producer;
+
+    public void commitConsumer(Properties producerProperties,
+                               BMap<String, BValue> producerConnector,
+                               Map<TopicPartition, OffsetAndMetadata> partitionToMetadataMap,
+                               String groupID) {
+
```
Comment (pamod) [inline component/src/main/java/org/ballerinalang/kafka/nativeimpl/producer/action/AbstractTransactionHandler.java]: Extra space ?

File: component/src/main/java/org/ballerinalang/kafka/util/KafkaUtils.java [hunk 16]
```diff
@@ -513,4 +533,20 @@ private static void addBooleanParamIfPresent(String paramName,
         return partitionList;
     }
 
+    public static List<BMap<String, BValue>> createPartitionList(Context context,
+                                                                   Collection<TopicPartition> partitions) {
+
+        List<BMap<String, BValue>> topicPartitionList = new ArrayList<>();
+        if (!partitions.isEmpty()) {
```
Comment (pamod) [inline component/src/main/java/org/ballerinalang/kafka/util/KafkaUtils.java]: If partition list is empty what would be the implication ?

Comment (ThisaruGuruge) [reply component/src/main/java/org/ballerinalang/kafka/util/KafkaUtils.java]: Then empty list will be returned.

**Pull request merged**